cupy/cuda/stream.pyx

```diff
@@ -178,32 +203,42 @@ class BaseStream(object):
         # This operator is implemented to compare the singleton instance
         # of null stream (Stream.null) can safely be compared with null
         # stream instance created by a user.
-        return self.ptr == other.ptr
+        if self.dev == -1 or other.dev == -1:
+            return self.ptr == other.ptr
+        else:
+            return (self.dev == other.dev and self.ptr == other.ptr)
 
     def __enter__(self):
         tls = _ThreadLocal.get()
-        if tls.prev_stream_ref_stack is None:
-            tls.prev_stream_ref_stack = []
-        prev_stream_ref = tls.get_current_stream_ref()
-        tls.prev_stream_ref_stack.append(prev_stream_ref)
+        cdef int dev = self.dev
```

Comment:
I think it's better to move `__enter__` (and `__exit__`) implementation to `tls` (like `tls.push_stream(self.ptr, self.dev)`) to aovid modifying TLS internal states in various places.